server/src/main/java/io/deephaven/server/arrow/ArrowFlightUtil.java

```diff
@@ -66,6 +71,166 @@ public class ArrowFlightUtil {
+            final SessionState session,
+            final TicketRouter ticketRouter,
+            final Flight.Ticket request,
+            final StreamObserver<InputStream> observer) {
+
+        final SessionState.ExportObject<BaseTable> export =
+                ticketRouter.resolve(session, request, "request");
+
+        final BarragePerformanceLog.SnapshotMetricsHelper metrics =
+                new BarragePerformanceLog.SnapshotMetricsHelper();
+
+        final long queueStartTm = System.nanoTime();
+        session.nonExport()
+                .require(export)
+                .onError(observer)
+                .submit(() -> {
+                    metrics.queueNanos = System.nanoTime() - queueStartTm;
+                    final BaseTable table = export.get();
+                    metrics.tableId = Integer.toHexString(System.identityHashCode(table));
+                    metrics.tableKey = BarragePerformanceLog.getKeyFor(table);
+
+                    // create an adapter for the response observer
+                    final StreamObserver<BarrageStreamGenerator.View> listener =
+                            ArrowModule.provideListenerAdapter().adapt(observer);
+
+                    // Send Schema wrapped in Message
+                    final FlatBufferBuilder builder = new FlatBufferBuilder();
+                    final int schemaOffset = BarrageUtil.makeSchemaPayload(builder, table.getDefinition(),
+                            table.getAttributes());
+                    builder.finish(MessageHelper.wrapInMessage(builder, schemaOffset,
+                            org.apache.arrow.flatbuf.MessageHeader.Schema));
+                    final ByteBuffer serializedMessage = builder.dataBuffer();
+
+                    // leverage the stream generator SchemaView constructor
+                    final BarrageStreamGenerator.SchemaView schemaView =
+                            new BarrageStreamGenerator.SchemaView(serializedMessage);
+
+                    // push the schema to the listener
+                    listener.onNext(schemaView);
+
+                    // shared code between `DoGet` and `BarrageSnapshotRequest`
+                    createAndSendSnapshot(table, null, null, false, DEFAULT_SNAPSHOT_DESER_OPTIONS, listener, metrics);
+
+                    listener.onCompleted();
+                });
+    }
+
+    private static void createAndSendSnapshot(BaseTable table, BitSet columns, RowSet viewport, boolean reverseViewport,
+            BarrageSnapshotOptions snapshotRequestOptions, StreamObserver<BarrageStreamGenerator.View> listener,
+            BarragePerformanceLog.SnapshotMetricsHelper metrics) {
+        // if the table is static or append-only and a full snapshot is requested, we can make and send multiple
+        // snapshots to save memory and operate more efficiently
+
+        if (viewport == null && (!table.isRefreshing() || table.isAddOnly())) {
```

Comment:
As Ryan mentioned the other day, this is `isAddOnly` and not `isAppendOnly`. We'll have to drop to static tables only.